src/asl.ts

```diff
@@ -2918,6 +2934,285 @@ export class ASL {
+      const rest = pattern.bindings.find((binding) => binding?.rest);
+      if (rest && isObjectBinding(pattern)) {
+        // TODO: validator
+        throw new SynthError(
+          ErrorCodes.StepFunctions_does_not_support_destructuring_object_with_rest
+        );
+      }
+
+      // run each of the assignments as a sequence of states, they should not rely on each other.
+      const assignments = pattern.bindings.map((binding, i) => {
+        if (!binding) {
+          return undefined;
+        }
+
+        const updatedValue: ASLGraph.Output = (() => {
+          if (isArrayBinding(pattern)) {
+            // const [a] = arr;
+            if (ASLGraph.isJsonPath(value)) {
+              return {
+                jsonPath: `${value.jsonPath}[${i}]`,
+              };
+            } else {
+              // const [a] = [1,2,3];
+              if (!Array.isArray(value.value)) {
+                throw new SynthError(
+                  ErrorCodes.Invalid_Input,
+                  "Expected array binding pattern to be on a reference path or array literal."
+                );
+              }
+              return {
+                value: value.value[i],
+                containsJsonPath: value.containsJsonPath,
+              };
+            }
+          } else {
+            // when `name` is a bindingPattern, propertyName should always be present.
+            const propertyNameExpr = binding.propertyName ?? binding.name;
+            const propertyName = isComputedPropertyNameExpr(propertyNameExpr)
+              ? evalToConstant(propertyNameExpr.expr)?.constant
+              : isIdentifier(propertyNameExpr)
+              ? propertyNameExpr.name
+              : isStringLiteralExpr(propertyNameExpr)
+              ? propertyNameExpr.value
+              : undefined;
+
+            if (!propertyName || typeof propertyName !== "string") {
+              throw new SynthError(
+                ErrorCodes.StepFunctions_property_names_must_be_constant
+              );
+            }
+
+            if (ASLGraph.isJsonPath(value)) {
+              return {
+                jsonPath: `${value.jsonPath}['${propertyName}']`,
+              };
+            } else if (value.value && typeof value.value === "object") {
+              return {
+                value: value.value[propertyName as keyof typeof value.value],
+                containsJsonPath: value.containsJsonPath,
+              };
+            } else {
+              throw new SynthError(
+                ErrorCodes.Invalid_Input,
+                "Expected object binding pattern to be on a reference path or object literal"
+              );
+            }
+          }
+        })();
+
+        // when there is a default value
+        if (binding.initializer) {
+          // if there is a default value, update the output to reflect it
+          const valueStatesWithDefault = this.applyDefaultValue(
+            updatedValue,
+            binding.initializer
+          );
```

Comment:
Does this set the value with a Condition?

Reply:
yeah